File: solutions/0/0/7/3/p0073_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 73: Counting Fractions in a Range [Level 3]. """
from __future__ import annotations

from solver.runners import runner


@runner.main
def solve(*args: str) -> str:
    """Iterative Farey neighbour traversal of (1/3, 1/2): the neighbour identity bc - ad = 1 gives
    the next denominator as max_d - ((max_d + prev_d) % d). Only denominators are tracked since the
    boundaries fix every numerator. O(answer)."""
    max_d = runner.parse_int(args[0])

    lower_denominator: int = 3
    upper_denominator: int = 2
    # Largest denominator of the form 2 + 3k not exceeding max_d: the fraction just above 1/3.
    d = upper_denominator + lower_denominator * ((max_d - upper_denominator) // lower_denominator)
    prev_d = lower_denominator
    count = 0
    # Walk upward, counting one fraction per step; stop at denominator 2 (the upper bound 1/2).
    while d != upper_denominator:
        count += 1
        prev_d, d = (d, max_d - (max_d + prev_d) % d)
    return str(count)


if __name__ == "__main__":
    raise SystemExit(solve())

File: solutions/0/0/7/3/p0073_s0.c
/* Solution to Euler Problem 73: Counting Fractions in a Range. */
#include "runner.h"

const char *solve(int argc, char *argv[]) {
    /* Iterative Farey neighbour traversal between 1/3 and 1/2: from the neighbour identity
       bc - ad = 1, the next denominator is d_next = D - ((D + prev_d) mod d). Only denominators
       are tracked (numerators are forced by the boundaries). O(answer) time. */
    static char _answer[32];
    int max_d = parse_int(argv[1]);
    int lower_denominator = 3;
    int upper_denominator = 2;
    /* Largest denominator of the form 2 + 3k not exceeding max_d: the fraction just above 1/3. */
    int d = upper_denominator + lower_denominator * ((max_d - upper_denominator) / lower_denominator);
    int prev_d = lower_denominator;
    long long count = 0;
    /* Walk upward, counting one fraction per step; stop at denominator 2 (the upper bound 1/2). */
    while (d != upper_denominator) {
        count++;
        int new_d = max_d - (max_d + prev_d) % d;
        prev_d = d;
        d = new_d;
    }
    { snprintf(_answer, sizeof _answer, "%lld", (long long)(count)); return _answer; }
}